File: backend/services/transfer_service.py
from decimal import Decimal
from sqlalchemy.orm import Session
from sqlalchemy import select
from fastapi import HTTPException
from backend.models.models import EventAction, EventStatus, User, Transaction
from backend.models.schemas import TransferStatus

class InsufficientFundsError(Exception):
    pass

class AccountNotFoundError(Exception):
    pass

class TransferService:
    def __init__(self, db: Session):
        self.db = db

    def execute(
        self,
        source_account_id: int,
        target_account_id: int,
        balance_user: int,
        amount: Decimal,
        description: str,
        type_movment: str
    ) -> Transaction:
        """
        Toda la operación ocurre dentro de una sola transacción.
        Si cualquier paso falla, SQLAlchemy hace rollback automático.
        """
        
        try:
            # Bloquear filas con SELECT FOR UPDATE 
            source = (
                self.db.execute(
                    select(User)
                    .where(User.id == source_account_id)
                    .with_for_update()         
                )
                .scalars()
                .first()
            )
            target = (
                self.db.execute(
                    select(User)
                    .where(User.id == target_account_id)
                    .with_for_update()
                )
                .scalars()
                .first()
            )

            # Validaciones 
            if not source:
                raise AccountNotFoundError(f"Cuenta origen {source_account_id} no existe")
            if not target:
                raise AccountNotFoundError(f"Cuenta destino {target_account_id} no existe")
            if amount <= 0:
                raise ValueError("El monto debe ser mayor a cero")
            if source.balance < amount:
                raise InsufficientFundsError(
                    f"Saldo insuficiente: disponible {source.balance}, requerido {amount}"
                )

            # Operación atómica: restar y sumar 
            source.balance -= amount
            target.balance += amount

            # Registrar la transferencia 
            transfer = Transaction(
                user_id = source_account_id,
                target_user_id = target_account_id,
                amount = amount,
                balance_after = balance_user - amount,
                description = description,
                type = type_movment
            )
            self.db.add(transfer)

            # Commit
            self.db.commit()
            self.db.refresh(transfer)
            return transfer

        except (InsufficientFundsError, AccountNotFoundError, ValueError):
            self.db.rollback()
            raise  

        except Exception as e:
            self.db.rollback()
            raise RuntimeError(f"Error inesperado en transferencia: {e}") from e

File: backend/src/services/transferService.js
const db = require('../db');
const { validateAccountNumber } = require('../helpers/validation');
const audit = require('./auditService');

async function transfer(sender, destinyNumber, amount, description) {
  if (!validateAccountNumber(destinyNumber)) {
    await audit.transferFailed(sender, `cuenta destino mal formada: ${destinyNumber}`);
    const e = new Error('Número de cuenta inválido'); e.status = 400; throw e;
  }

  return db.transaction(async (trx) => {
    const src = await trx('users').where({ id: sender.id }).forUpdate().first();
    const dst = await trx('users').where({ number: destinyNumber }).forUpdate().first();

    if (!dst) {
      await audit.transferFailed(sender, `cuenta destino inexistente: ${destinyNumber}`);
      const e = new Error('Num typed doesnt correspond to any user'); e.status = 404; throw e;
    }
    if (src.balance < amount) {
      await audit.transferFailed(sender, `saldo insuficiente: monto ${amount}, saldo ${src.balance}`);
      const e = new Error('Insufficient balance'); e.status = 400; throw e;
    }

    await trx('users').where({ id: src.id }).update({ balance: src.balance - amount });
    await trx('users').where({ id: dst.id }).update({ balance: dst.balance + amount });

    const [txId] = await trx('transactions').insert({
      user_id: src.id,
      target_user_id: dst.id,
      type: 'transfer',
      amount,
      balance_after: src.balance - amount,
      description: description || `transfer a ${destinyNumber}`,
      date: new Date(),
    });

    const tx = await trx('transactions').where({ id: txId }).first();
    await audit.transferApproved(sender, `transfer de ${amount} - balance: ${src.balance - amount}`);
    return tx;
  });
}

module.exports = { transfer };
